### PR TITLE
Expose retry_budget configuration in DestinationRule ConnectionPoolSettings

### PR DESCRIPTION
**Pull request by ericdbishop**
**Please provide a description of this PR:**

Attempting to expose the Envoy CircuitBreaker retry_budget threshold within DestinationRule configuration. The only alternate way to maintain retry_budget configurations at the moment is through complex EnvoyFilter’s, which becomes very difficult to maintain across hundreds of separate microservices within the mesh & requires automation to generate the configs.

We are aware that there is concern about forwards compatibility with the work that @mikemorris is leading for the Kubernetes Gateway API in order to expose retry_budget configuration within HTTPRoute. But we don’t currently have plans to adopt Gateway API in the foreseeable future, and exposing retry_budget alongside max_retries within the DestinationRule seems to make the most sense in the context of Istio based off of [Envoy CircuitBreaker configuration](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/circuit_breaker.proto#config-cluster-v3-circuitbreakers-thresholds).

From our own testing, we have found that a dynamic retry budget performs far better than configuring a static value for max retries. Would be happy to discuss this more, but we feel that this is an important feature based off of the other issues and PR’s that have been submitted for exposing this configuration, as well as what we’ve found through testing. Looking forward to feedback as this is a high priority feature for us. cc @tejokumar

api: https://github.com/istio/api/pull/3249

Credit to @quraynai as I used [their previous work in this area](https://github.com/istio/istio/pull/47094) as a reference.
Issue: https://github.com/istio/istio/issues/27419

### CODE REVIEW

**Comment by ericdbishop** [comment]
Hi @craigbox @howardjohn @mikemorris, I was hoping to get a discussion going on this topic. Currently, manually generating retry budget EnvoyFilter's is creating a complex challenge for us at build time. 

In order to utilize retry budgets, when generating application manifests for any given service, we must also generate an individual EnvoyFilter configuration within each namespace of every client application. This creates an additional 200+ lines of configuration per client app of a service, across hundreds of microservices. As this functionality is critical for us, I'd like to see if we could revisit the topic of exposing the `retry_budget` threshold within DestinationRule.

**Comment by howardjohn** [comment]
I don't feel comfortable making any Istio API changes until https://github.com/kubernetes-sigs/gateway-api/pull/3199 is ironed out. Even if you don't plan to use the Gateway API, the decisions there will still impact us. For example, there has been serious consideration to making a retry budget at the HTTPRoute level in GW-API, and implementing equivalent into Envoy Route.

If that were to happen, we would definitely want to add retry_budget in Istio into VirtualService, not DestinationRule. Which is actually my preferred end state anyways, as it is confusing to users to have retry policies artificially split between two objects.

**Comment by ericdbishop** [comment]
Thanks for the feedback @howardjohn. That is understandable that the priority is maintaining Gateway API compatibility. Though I'm wondering how this may look within VirtualService when specifying multiple destinations on the route? Each destination would need a specific retry budget, correct? Wasn't aware that the idea was to contribute this directly into Envoy Route.

**Comment by ericdbishop** [comment]
@mikemorris, my question in that other PR wasn't so much about being able to set a default across all services, but the fact that setting a retry_budget for all services by generating EnvoyFilter's creates much more complexity when generating manifests/configurations to be applied in the cluster. If `retry_budget` was an additional field in the [DestinationRule ConnectionPoolSettings](https://istio.io/latest/docs/reference/config/networking/destination-rule/#ConnectionPoolSettings-HTTPSettings), that would let us add in a sane default for all services at build time, without thousands of lines of additional EnvoyFilter configurations.

Setting a retry_budget policy for the threshold of concurrent retries at the destination service was what we were going for, as that is where we are currently setting a static threshold for parallel retries at the destination through DestinationRule ConnectionPoolSettings.

[HTTPRetry in the VirtualService](https://istio.io/latest/docs/reference/config/networking/virtual-service/#HTTPRetry) in Istio is used to configure the number of retry attempts from the client for a route. I find that setting a threshold on the client side for parallel/concurrent retries to be slightly confusing from a UX perspective for this reason as well, since **concurrent attempts** are different than the number of **total attempts**. And then you would have the added complexity of setting thresholds on each route. Hopefully that clears up my concern a bit, but curious to see where this goes and let me know if I can be of any help.

**Comment by ericdbishop** [comment]
https://github.com/istio/api/pull/3249#issuecomment-2243148466

**Comment by alexander-magon** [comment]
Hi @craigbox @howardjohn @mikemorris, can we re-establish conversation around this PR?
retry_budget is quite important feature and both PRs are ready to review, current and [istio/api#3249](https://github.com/istio/api/pull/3249)

**Comment by alexander-magon** [comment]
Hi @craigbox @howardjohn @mikemorris, @keithmattix can you please take a look or help to find reviewers for this [istio/API](https://github.com/istio/api/pull/3249) PR as it is dependency for current

File: pilot/pkg/networking/core/cluster_traffic_policy.go [hunk 1]
```diff
@@ -348,6 +358,14 @@ func getDefaultCircuitBreakerThresholds() *cluster.CircuitBreakers_Thresholds {
 	}
 }
 
+// getDefaultCircuitBreakerThresholdsRetryBudget returns a copy of the default circuit breaker thresholds retry budget for the given traffic direction.
+func getDefaultCircuitBreakerThresholdsRetryBudget() *cluster.CircuitBreakers_Thresholds_RetryBudget {
+	return &cluster.CircuitBreakers_Thresholds_RetryBudget{
+		BudgetPercent: &xdstype.Percent{Value: math.MaxFloat64},
+		MinRetryConcurrency: &wrapperspb.UInt32Value{Value: math.MaxUint32},
```
Comment (ajvengo) [inline pilot/pkg/networking/core/cluster_traffic_policy.go:365]: Why default values here not the same as in the defaults reported in the comments [here](https://github.com/istio/api/pull/3249/files#diff-12899beb695152c851fb77fb9a8983897fe1c9c17eaf7665b146b8d55a9a5c18R663-R673)?
```suggestion
		BudgetPercent: &xdstype.Percent{Value: 20},
		MinRetryConcurrency: &wrapperspb.UInt32Value{Value: 3},
```

**Pull request closed without merging**